mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,220 @@
+    * @param maybeCredentials The session token if required.
+    * @return The [[HttpRequest]] with proper headers and body.
+    */
+  def createPostRequest(bytes: Array[Byte], maybeCredentials: Option[HttpCredentials]): HttpRequest = {
+    println(s"Body: ${bytes.map(_.toChar).mkString}")
+    HttpRequest(
+      HttpMethods.POST,
+      uri = Uri(s"$url/api/v1/scheduler"),
+      entity = HttpEntity(MesosClient.ProtobufMediaType, bytes),
+      headers = MesosClient.MesosStreamIdHeader(streamId) :: maybeCredentials.map(Authorization(_)).toList
+    )
+  }
+
+  // Fail and trigger restart if the call was unauthorized.
+  def handleRejection(response: Try[HttpResponse]): Try[HttpResponse] = {
+    response match {
+      case Success(r) if r.status == StatusCodes.Unauthorized =>
+        throw new IllegalStateException("Session token expired.")
+      case id => id
+    }
+  }
+
+  /** @return A flow that transforms serialized Mesos calls to proper HTTP requests. */
+  def post(connection: Flow[HttpRequest, Try[HttpResponse], NotUsed]): Flow[Array[Byte], Try[HttpResponse], NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        RestartFlow.withBackoff(1.second, 1.second, 1, 3)(() =>
+          Flow.fromGraph(SessionFlow(credentialsProvider, createPostRequest)).via(connection).map(handleRejection))
+      case None =>
+        Flow[Array[Byte]].map(createPostRequest(_, None)).via(connection)
+    }
+
+  def postSimple(implicit system: ActorSystem): Flow[Array[Byte], Try[HttpResponse], NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        implicit val askTimeout = util.Timeout(20.seconds)
+        val sessionActor = system.actorOf(SessionActor.props(credentialsProvider, createPostRequest))
+        Flow[Array[Byte]].ask[HttpResponse](1)(sessionActor).map(Success(_))
+      case None =>
+        Flow[Array[Byte]].map(createPostRequest(_, None)).mapAsync(1)(Http().singleRequest(_)).map(Success(_))
+    }
+
+  /** @return The connection pool for this session. */
+  def connectionPool(implicit system: ActorSystem): Flow[HttpRequest, Try[HttpResponse], NotUsed] = {
+    // Disable pipelining.
+    val poolSettings = ConnectionPoolSettings("").withMaxConnections(1).withPipeliningLimit(1)
+    if (isSecured) {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().cachedHostConnectionPoolHttps(host = url.getHost, port = port, settings = poolSettings))
+        .map(_._1)
+    } else {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().cachedHostConnectionPool(host = url.getHost, port = port, settings = poolSettings))
+        .map(_._1)
+    }
+  }
+}
+
+object SessionActor {
+  def props(
+      credentialsProvider: CredentialsProvider,
+      requestFactory: (Array[Byte], Option[HttpCredentials]) => HttpRequest): Props = {
+    Props(new SessionActor(credentialsProvider, requestFactory))
+  }
+}
+
+class SessionActor(
```

Comment:
This is an new proposal following the Colin Breck's [posts](https://blog.colinbreck.com/integrating-akka-streams-and-akka-actors-part-ii/). I have to admit that he is right: Handling state and retries is much simpler in an actor. Please compare `Session#post` with `Session#postSimple`. `Flow.ask` will also yield backpressure and since we are doing single sequential HTTP requests anyways we would not benefit from a connection pool anyways. Strict cluster send a `Connection: close` header so we have to reestablish the connection for each Mesos call anyways.